### PR TITLE
examples: fix C example servers retry

### LINKED ISSUE #566: C version examples [server] and [client] connection failed
I am trying c version examples in quiche/examples and I have some trouble using them.
Use master branch, build these examples by `make` in quiche/examples
Just changed one line in client.c:
```c
-    quiche_config *config = quiche_config_new(0xbabababa);
+    quiche_config *config = quiche_config_new(QUICHE_PROTOCOL_VERSION);
```

Here are logs:
server
``` bash
~/.../quiche/examples >>> ./server 127.0.0.1 1234                                           
stateless retry
sent 93 bytes
recv would block
new connection
quiche: bed00effe987dca943b3a186386830a7 rx pkt Initial version=ff00001d dcid=c4c07e8f8588b0f65febb00f9c62d991 scid=57fb5bcfc3946bf7cd5800bb49a75417 token=7175696368650200ed7f7f0000010000000000000000c4c07e8f8588b0f65febb00f9c62d991 len=1120 pn=1
quiche: bed00effe987dca943b3a186386830a7 rx frm CRYPTO off=0 len=512
quiche::tls: checking peer ALPN Ok("hq-29") against Ok("hq-29")
quiche::tls: bed00effe987dca943b3a186386830a7 write message lvl=Initial len=90
quiche::tls: bed00effe987dca943b3a186386830a7 set write secret lvl=Handshake
quiche::tls: bed00effe987dca943b3a186386830a7 write message lvl=Handshake len=1178
quiche::tls: bed00effe987dca943b3a186386830a7 set write secret lvl=OneRTT
quiche::tls: bed00effe987dca943b3a186386830a7 set read secret lvl=Handshake
quiche: bed00effe987dca943b3a186386830a7 rx frm PADDING len=587
recv 1200 bytes
recv would block
quiche: bed00effe987dca943b3a186386830a7 tx pkt Initial version=ff00001d dcid=57fb5bcfc3946bf7cd5800bb49a75417 scid=bed00effe987dca943b3a186386830a7 len=116 pn=0
quiche: bed00effe987dca943b3a186386830a7 tx frm ACK delay=1118 blocks=[1..1]
quiche: bed00effe987dca943b3a186386830a7 tx frm CRYPTO off=0 len=90
quiche::recovery: bed00effe987dca943b3a186386830a7 timer=999.850593ms latest_rtt=0ns srtt=None min_rtt=0ns rttvar=0ns loss_time=[None, None, None] loss_probes=[0, 0, 0] cwnd=14520 ssthresh=18446744073709551615 bytes_in_flight=159 app_limited=true congestion_recovery_start_time=None delivered=0 delivered_time=162.421µs recent_delivered_packet_sent_time=163.121µs app_limited_at_pkt=0  hystart=window_end=None last_round_min_rtt=None current_round_min_rtt=None rtt_sample_count=0 lss_start_time=None  
sent 159 bytes
quiche: bed00effe987dca943b3a186386830a7 tx pkt Handshake version=ff00001d dcid=57fb5bcfc3946bf7cd5800bb49a75417 scid=bed00effe987dca943b3a186386830a7 len=1154 pn=0
quiche: bed00effe987dca943b3a186386830a7 tx frm CRYPTO off=0 len=1134
quiche::recovery: bed00effe987dca943b3a186386830a7 timer=999.711623ms latest_rtt=0ns srtt=None min_rtt=0ns rttvar=0ns loss_time=[None, None, None] loss_probes=[0, 0, 0] cwnd=14520 ssthresh=18446744073709551615 bytes_in_flight=1355 app_limited=true congestion_recovery_start_time=None delivered=0 delivered_time=296.695µs recent_delivered_packet_sent_time=297.22µs app_limited_at_pkt=0  hystart=window_end=None last_round_min_rtt=None current_round_min_rtt=None rtt_sample_count=0 lss_start_time=None  
sent 1196 bytes
quiche: bed00effe987dca943b3a186386830a7 tx pkt Handshake version=ff00001d dcid=57fb5bcfc3946bf7cd5800bb49a75417 scid=bed00effe987dca943b3a186386830a7 len=64 pn=1
quiche: bed00effe987dca943b3a186386830a7 tx frm CRYPTO off=1134 len=44
quiche::recovery: bed00effe987dca943b3a186386830a7 timer=999.600422ms latest_rtt=0ns srtt=None min_rtt=0ns rttvar=0ns loss_time=[None, None, None] loss_probes=[0, 0, 0] cwnd=14520 ssthresh=18446744073709551615 bytes_in_flight=1461 app_limited=true congestion_recovery_start_time=None delivered=0 delivered_time=410.724µs recent_delivered_packet_sent_time=411.244µs app_limited_at_pkt=0  hystart=window_end=None last_round_min_rtt=None current_round_min_rtt=None rtt_sample_count=0 lss_start_time=None  
sent 106 bytes
done writing
quiche: bed00effe987dca943b3a186386830a7 rx pkt Handshake version=ff00001d dcid=bed00effe987dca943b3a186386830a7 scid=57fb5bcfc3946bf7cd5800bb49a75417 len=21 pn=0
quiche: bed00effe987dca943b3a186386830a7 rx frm CONNECTION_CLOSE err=8 frame=0 reason=[]
quiche: bed00effe987dca943b3a186386830a7 dropped epoch 0 state
recv 61 bytes
recv would block
done writing
quiche: bed00effe987dca943b3a186386830a7 draining timeout expired
timeout
done writing
connection closed, recv=2 sent=3 lost=0 rtt=500000000ns cwnd=14520
```
client
```bash
~/.../quiche/examples >>> ./client 127.0.0.1 1234             
quiche::tls: 57fb5bcfc3946bf7cd5800bb49a75417 write message lvl=Initial len=512
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 tx pkt Initial version=ff00001d dcid=c4c07e8f8588b0f65febb00f9c62d991 scid=57fb5bcfc3946bf7cd5800bb49a75417 len=1157 pn=0
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 tx frm CRYPTO off=0 len=512
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 tx frm PADDING len=625
quiche::recovery: 57fb5bcfc3946bf7cd5800bb49a75417 timer=999.544142ms latest_rtt=0ns srtt=None min_rtt=0ns rttvar=0ns loss_time=[None, None, None] loss_probes=[0, 0, 0] cwnd=14520 ssthresh=18446744073709551615 bytes_in_flight=1200 app_limited=true congestion_recovery_start_time=None delivered=0 delivered_time=472.271µs recent_delivered_packet_sent_time=472.809µs app_limited_at_pkt=0  hystart=window_end=None last_round_min_rtt=None current_round_min_rtt=None rtt_sample_count=0 lss_start_time=None  
sent 1200 bytes
done writing
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 rx pkt Retry version=ff00001d dcid=57fb5bcfc3946bf7cd5800bb49a75417 scid=c4c07e8f8588b0f65febb00f9c62d991 token=7175696368650200ed7f7f0000010000000000000000c4c07e8f8588b0f65febb00f9c62d991
recv 93 bytes
recv would block
done reading
quiche::tls: 57fb5bcfc3946bf7cd5800bb49a75417 write message lvl=Initial len=512
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 tx pkt Initial version=ff00001d dcid=c4c07e8f8588b0f65febb00f9c62d991 scid=57fb5bcfc3946bf7cd5800bb49a75417 token=7175696368650200ed7f7f0000010000000000000000c4c07e8f8588b0f65febb00f9c62d991 len=1119 pn=1
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 tx frm CRYPTO off=0 len=512
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 tx frm PADDING len=587
quiche::recovery: 57fb5bcfc3946bf7cd5800bb49a75417 timer=999.582502ms latest_rtt=0ns srtt=None min_rtt=0ns rttvar=0ns loss_time=[None, None, None] loss_probes=[0, 0, 0] cwnd=14520 ssthresh=18446744073709551615 bytes_in_flight=1200 app_limited=true congestion_recovery_start_time=None delivered=0 delivered_time=1.48449ms recent_delivered_packet_sent_time=1.48498ms app_limited_at_pkt=0  hystart=window_end=None last_round_min_rtt=None current_round_min_rtt=None rtt_sample_count=0 lss_start_time=None  
sent 1200 bytes
done writing
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 rx pkt Initial version=ff00001d dcid=57fb5bcfc3946bf7cd5800bb49a75417 scid=bed00effe987dca943b3a186386830a7 token= len=117 pn=0
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 rx frm ACK delay=1118 blocks=[1..1]
quiche::recovery: 57fb5bcfc3946bf7cd5800bb49a75417 packet newly acked 1
quiche::recovery: 57fb5bcfc3946bf7cd5800bb49a75417 timer=none latest_rtt=10.321509ms srtt=Some(10.321509ms) min_rtt=10.321509ms rttvar=5.160754ms loss_time=[None, None, None] loss_probes=[0, 0, 0] cwnd=14520 ssthresh=18446744073709551615 bytes_in_flight=0 app_limited=true congestion_recovery_start_time=None delivered=1200 delivered_time=157.869µs recent_delivered_packet_sent_time=11.539471ms app_limited_at_pkt=0  hystart=window_end=None last_round_min_rtt=None current_round_min_rtt=None rtt_sample_count=0 lss_start_time=None  
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 rx frm CRYPTO off=0 len=90
quiche::tls: 57fb5bcfc3946bf7cd5800bb49a75417 set write secret lvl=Handshake
quiche::tls: 57fb5bcfc3946bf7cd5800bb49a75417 set read secret lvl=Handshake
recv 159 bytes
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 rx pkt Handshake version=ff00001d dcid=57fb5bcfc3946bf7cd5800bb49a75417 scid=bed00effe987dca943b3a186386830a7 len=1155 pn=0
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 rx frm CRYPTO off=0 len=1134
failed to process packet
recv 106 bytes
recv would block
done reading
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 tx pkt Handshake version=ff00001d dcid=bed00effe987dca943b3a186386830a7 scid=57fb5bcfc3946bf7cd5800bb49a75417 len=20 pn=0
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 tx frm CONNECTION_CLOSE err=8 frame=0 reason=[]
quiche::recovery: 57fb5bcfc3946bf7cd5800bb49a75417 timer=44.094416ms latest_rtt=10.321509ms srtt=Some(10.321509ms) min_rtt=10.321509ms rttvar=5.160754ms loss_time=[None, None, None] loss_probes=[0, 0, 0] cwnd=14520 ssthresh=18446744073709551615 bytes_in_flight=61 app_limited=true congestion_recovery_start_time=None delivered=1200 delivered_time=1.560008ms recent_delivered_packet_sent_time=12.941538ms app_limited_at_pkt=0  hystart=window_end=None last_round_min_rtt=None current_round_min_rtt=None rtt_sample_count=0 lss_start_time=None  
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 dropped epoch 0 state
sent 61 bytes
done writing
timeout
done writing
quiche: 57fb5bcfc3946bf7cd5800bb49a75417 draining timeout expired
timeout
done writing
connection closed, recv=1 sent=3 lost=0 rtt=10321509ns
```

Thanks for your help!

### PR DESCRIPTION
**Pull request by LPardue**
fixes #566 

Clients connecting to these servers fail when checking `retry_source_connection_id`. This change aligns the C examples more closely with the Rust examples by storing a connection object when we do a stateless retry. We only create a full QUIC connection once the client present a subsequent valid token.

### CODE REVIEW

File: examples/http3-server.c [hunk 1]
```diff
@@ -175,11 +175,20 @@ static struct conn_io *create_conn(uint8_t *odcid, size_t odcid_len) {
...
 
+    conn_io->conn = NULL;
+
+    HASH_ADD(hh, conns->h, cid, LOCAL_CONN_ID_LEN, conn_io);
+
+    return conn_io;
+}
+
+static ssize_t accept_conn(struct conn_io *conn_io, uint8_t *odcid, size_t odcid_len) {
+
```
Comment (ghedo) [inline examples/http3-server.c:186]:
```suggestion
```

File: examples/http3-server.c [hunk 2]
```diff
@@ -205,6 +212,7 @@ static int for_each_header(uint8_t *name, size_t name_len,
 }
 
 static void recv_cb(EV_P_ ev_io *w, int revents) {
+    fprintf(stderr, "receiving\n");
```
Comment (ghedo) [inline examples/http3-server.c:215]:
```suggestion
```

File: examples/http3-server.c [hunk 3]
```diff
@@ -432,10 +451,16 @@ static void recv_cb(EV_P_ ev_io *w, int revents) {
             free(conn_io);
         }
     }
+
+    fprintf(stderr, "left HASH_ITER\n");
```
Comment (ghedo) [inline examples/http3-server.c:455]:
```suggestion
```

File: examples/http3-server.c [hunk 4]
```diff
@@ -432,10 +451,16 @@ static void recv_cb(EV_P_ ev_io *w, int revents) {
             free(conn_io);
         }
     }
+
```
Comment (ghedo) [inline examples/http3-server.c:454]:
```suggestion
```

**Comment by LPardue** [comment]
> The idea of "stateless" retry is that we don't keep state on the server, but it seems this removes the stateless part as it creates the connection object in advance, which is not quite what we do in the Rust examples.

> I also don't think I understand what the original problem is TBH, though it's definitely possible that the C examples are out-of-sync with the Rust ones.

Good point. I wonder then if this is caused by the current code failing to use the correct values when the validate_token call passes, it's kinda hard to follow what values for various CIDs are used before and after the token is parsed. We might be able to fix the core issue in a simpler way.

**Comment by ghedo** [comment]
Replaced by https://github.com/cloudflare/quiche/commit/6d7bab4468a59ff1e48afede6cab1d696c363026.
